assets/tests/TaskSchemaValidationService.js

```diff
@@ -0,0 +1,225 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for validating data put into a task schema
+ * within a question.
+ */
+
+tie.factory('TaskSchemaValidationService', ['SYSTEM_CODE',
+  function(SYSTEM_CODE) {
+    var AUXILIARY_CODE_CLASS_NAME = 'AuxiliaryCode.';
+    var SYSTEM_CODE_CLASS_NAME = 'System.';
+    // TODO(eyurko): Update this once we support nonlinear runtimes.
+    var ALLOWED_RUNTIMES = ['linear'];
+    var _checkIfFunctionExistsInCode = function(functionName, code) {
+      var functionDefinition = 'def ' + functionName;
+      return code.indexOf(functionDefinition) !== -1;
+    };
+    var _checkIfFunctionExistsInClass = function(functionName, 
+      className, code) {
+      if (functionName.startsWith(className)) {
+        var strippedFunctionName = functionName.substring(
+          className.length);
+        return _checkIfFunctionExistsInCode(
+          strippedFunctionName, code);
+      }
+      return false;
+    };
+    var taskSchemaValidator = {
+      _starterCode: '',
```

Comment:
Ah, I see now where the "this" confusion was coming from -- I was wondering (while we were chatting) why it was an issue. Can you declare these as private vars in the service (at the same level as the functions above like _checkIfFunctionExistsInCode)? Then you would just refer to them without needing "this" or anything complicated, and you can just return this object the normal way.